File: 0086-partition-list/0086-partition-list.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val = 0, _next = nil)
#     @val = val
#     @next = _next
#   end
# end
# @param {ListNode} head
# @param {Integer} x
# @return {ListNode}
def partition(head, x)
  return head if !head
  left, right = ListNode.new, ListNode.new
  curr, left_tail, right_tail = head, left, right
  next_node = nil
  while curr
    if curr.val < x
      left_tail.next = curr
      left_tail = left_tail.next
    else
      right_tail.next = curr
      right_tail = right_tail.next
    end
    next_node = curr.next
    curr.next = nil
    curr = next_node
  end
  left_tail.next = right.next
  left.next
end

File: 0086-partition-list/0086-partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
  if head == nil {
    return head
  }
  left, right, curr := &ListNode{}, &ListNode{}, head
  leftTail, rightTail := left, right
  var nextNode *ListNode
  for curr != nil {
    if curr.Val < x {
      leftTail.Next = curr
      leftTail = leftTail.Next
    } else {
      rightTail.Next = curr
      rightTail = rightTail.Next
    }
    nextNode = curr.Next
    curr.Next = nil
    curr = nextNode
  }
  leftTail.Next = right.Next
  return left.Next
}